### PR TITLE
feat: specify id when provisional create canister

### PR DESCRIPTION
**Pull request by lwshang**
Close SDK-803

# Description

When create canister on non-mainnet replica, you can now specify the canister ID.

`dfx canister create <CANISTER_NAME> --specified-id <PRINCIPAL>`

`dfx deploy <CANISTER_NAME> --specified-id <PRINCIPAL>`

**Other changes**

- Did some re-org of the `Cargo.toml` files. The frontend canister is updated accordingly.
- Update ic-ref because the feature requires corresponding ref impl to pass e2e tests.

# How Has This Been Tested?

e2e

# Checklist:

- [x] The title of this PR complies with [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/).
- [x] I have edited the CHANGELOG accordingly.
- [x] I have made corresponding changes to the documentation.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,6 +4,14 @@
 
 ## DFX
 
+### feat: specify id when provisional create canister
+
+When create canister on non-mainnet replica, you can now specify the canister ID.
```
Comment (ericswanson-dfinity) [inline CHANGELOG.md:9]:
```suggestion
When creating a canister on non-mainnet replica, you can now specify the canister ID.
```

File: src/dfx/src/commands/canister/create.rs [hunk 2]
```diff
@@ -196,10 +207,12 @@ pub async fn exec(
                 .with_context(|| {
                     format!("Failed to read freezing threshold of {}.", canister_name)
                 })?;
+                // TODO: create for pull canisters with their mainnet ID
```
Comment (ericswanson-dfinity) [inline src/dfx/src/commands/canister/create.rs:210]: Is there a JIRA ticket that tracks this TODO? If so, please link to it in this comment. If not, please create one.

File: src/dfx/src/lib/operations/canister/create_canister.rs [hunk 3]
```diff
@@ -91,6 +96,7 @@ pub async fn create_canister(
                         .with_optional_freezing_threshold(settings.freezing_threshold)
                         .call_and_wait()
                         .await;
+                    dbg!(&res);
```
Comment (ericswanson-dfinity) [inline src/dfx/src/lib/operations/canister/create_canister.rs:99]:
```suggestion
```

**Review by ericswanson-dfinity — CHANGES_REQUESTED** [review]

**Review by ericswanson-dfinity — APPROVED** [review]

**Comment by LiveDuo** [comment]
> Any plans to support configuring a CID in dfx.json?

This is very welcome for setting up ledger, cmc or II locally. Currently it's quite a hassle.

**Comment by lwshang** [comment]
> setting up ledger, cmc or II locally.
@LiveDuo Could you give `dfx nns install` a try? It set system canisters locally.

"configuring a CID in dfx.json" is on our plan. It will firstly be in the upcoming `dfx pull` feature. For more general usage, we need design an intuitive workflow first.

**Comment by LiveDuo** [comment]
> > setting up ledger, cmc or II locally.
> 
> @LiveDuo Could you give `dfx nns install` a try? It set system canisters locally.
> 
> "configuring a CID in dfx.json" is on our plan. It will firstly be in the upcoming `dfx pull` feature. For more general usage, we need design an intuitive workflow first.

When it deploys the ledger canister who gets the initial distribution of the ICP?

PS: out of curiosity what's the `dfx pull` command for?

**Comment by lwshang** [comment]
You can specify the accounts as below:
https://github.com/dfinity/sdk/blob/900f9f016fb0b3622af536ef2b5eaffc60c4a14f/e2e/tests-dfx/ledger.bash#L17

`dfx pull` will help you to pull canisters (and their dependencies) from mainnet. Then you can install them locally at the same canister ID. It will simplify the procedure to integrate with other canisters (ledger, cmc, "some DEX", etc).
Check this file https://github.com/dfinity/sdk/blob/master/e2e/assets/pull/dfx.json to get an impression on how to specify dependency canisters to be pulled from mainnet.

**Comment by LiveDuo** [comment]
Thanks.

Ideally `dfx nns install` should be a flag in dfx.json too. I expect some projects using the nns canisters needing it all the time.

The `dfx pull` command should be great for e2e tests. Glad to learned about it.

**Pull request merged**